sklearn/compose/_column_transformer.py

```diff
@@ -0,0 +1,591 @@
+            Parameter names mapped to their values.
+        """
+        return self._get_params('_transformers', deep=deep)
+
+    def set_params(self, **kwargs):
+        """Set the parameters of this estimator.
+
+        Valid parameter keys can be listed with ``get_params()``.
+
+        Returns
+        -------
+        self
+        """
+        self._set_params('_transformers', **kwargs)
+        return self
+
+    def _iter(self, X=None, fitted=False, replace_strings=False):
+        """Generate (name, trans, column, weight) tuples
+        """
+        if fitted:
+            transformers = self.transformers_
+        else:
+            transformers = self.transformers
+        get_weight = (self.transformer_weights or {}).get
+
+        for name, trans, column in transformers:
+            if X is None:
+                sub = X
+            else:
+                sub = _get_column(X, column)
+
+            if replace_strings:
+                # replace 'passthrough' with identity transformer and
+                # skip in case of 'drop'
+                if trans == 'passthrough':
+                    trans = FunctionTransformer(
+                        validate=False, accept_sparse=True,
+                        check_inverse=False)
+                elif trans == 'drop':
+                    continue
+
+            yield (name, trans, sub, get_weight(name))
+
+    def _validate_transformers(self):
+        names, transformers, _, _ = zip(*self._iter())
+
+        # validate names
+        self._validate_names(names)
+
+        # validate estimators
+        for t in transformers:
+            if t in ('drop', 'passthrough'):
+                continue
+            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
+                    hasattr(t, "transform")):
+                raise TypeError("All estimators should implement fit and "
+                                "transform, or can be 'drop' or 'passthrough' "
+                                "specifiers. '%s' (type %s) doesn't." %
+                                (t, type(t)))
+
+    def _validate_unspecified(self, X):
+        """Generate list of passthrough columns for 'unspecified' case.
+        """
+        if self.unspecified not in ('drop', 'passthrough'):
+            raise ValueError(
+                "The unspecified keywords needs to be one of 'drop' or "
+                "'passthrough'. {0:r} was passed instead")
+
+        n_columns = X.shape[1]
+
+        if self.unspecified == 'passthrough':
+            cols = []
+            for _, _, columns in self.transformers:
+                cols.extend(_get_column_indices(X, columns))
+            self._passthrough = sorted(list(set(range(n_columns)) - set(cols)))
+        else:
+            self._passthrough = None
+
+    @property
+    def named_transformers_(self):
+        """Access the fitted transformer by name.
+
+        Read-only attribute to access any transformer by given name.
```

Comment:
can/should we force this to be read-only? Probably documenting is fine?

Reply:
This is a similar approach as eg with Pipeline, where it is only documented to be read-only. 

I also don't think there is an easy way to actually make it read-only, since even if Bunch itself is readonly, the attributes refer to the transformers, which are still mutable itself.